File: openvino-cplus/misc_classification_onnx_wthlabels.cpp
#include <stdio.h>
#include <math.h>

#include <iostream>
#include <fstream>
#include <vector>
#include <memory>
#include <functional>
#include <unistd.h>
#include "openvino/openvino.hpp"
#include <opencv2/opencv.hpp>
#include <boost/algorithm/string.hpp>

#include "samples/classification_results.h"
#include "format_reader_ptr.h"

long long get_nanoseconds() {
    // 定义一个timespec结构体，用于存储时间信息
    struct timespec ts;
    // 调用clock_gettime函数，获取当前时间，存储在ts中
    clock_gettime(CLOCK_MONOTONIC, &ts);
    // 返回ts中的秒数乘以10的9次方，加上纳秒数
    return ts.tv_sec * 1000000000LL + ts.tv_nsec;
}

void openvino_classifier(std::vector<std::string> image_file_names)
{
    std::vector<std::string> labels;
    std::string line;
    std::ifstream ifs_labels("labels-imagenet.txt");
    while(getline(ifs_labels, line)) {
        labels.push_back(line);
    }
    ifs_labels.close();

    size_t input_batch_size = 1;
    //size_t input_batch_size = image_file_names.size();

    size_t input_channel_size = 3;
    ov::Core core;
    //auto model = core.compile_model("classifier.onnx");
    //auto model = core.compile_model("resnet50.onnx");
    auto model = core.compile_model("resnet50-dynamic.onnx");
    auto iq = model.create_infer_request();
    auto output = iq.get_output_tensor(0);
    //clock_t start1, start2, stop;
    long long ns_start1, ns_start2, ns_stop;
    //start1 = clock();
    ns_start1 = get_nanoseconds();
    //std::cout << "start1：" << start1 << std::endl;
    /*
    size_t input_height_size = 224;
    size_t input_width_size = 224;
    */
    /*
    auto image1 = cv::imread(image_file_names[0]);
    size_t input_height_size = image1.rows;
    size_t input_width_size = image1.cols;
    std::cout << "image1.rows: " << image1.rows << std::endl;
    std::cout << "image1.cols: " << image1.cols << std::endl;
    */
    /*
    input.set_shape({input_batch_size, input_channel_size, input_width_size, input_height_size});
    float* input_data_host = input.data<float>();
    */

    for (size_t i = 0; i < image_file_names.size(); ++i) {
        const auto& image_file_name = image_file_names[i];
        auto image = cv::imread(image_file_name);
        size_t input_height_size = image.rows;
        size_t input_width_size = image.cols;
        std::cout << "image1.rows: " << image.rows << std::endl;
        std::cout << "image1.cols: " << image.cols << std::endl;
        auto input = iq.get_input_tensor(0);
        input.set_shape({input_batch_size, input_channel_size, input_width_size, input_height_size});
        float* input_data_host = input.data<float>();
        float mean[] = {0.406, 0.456, 0.485};
        float std[] = {0.225, 0.224, 0.229};

        //cv::resize(image, image, cv::Size(input_width_size, input_height_size));
        int image_area = image.cols * image.rows;
        unsigned char* pimage = image.data;
        float* phost_b = input_data_host + image_area * 0;
        float* phost_g = input_data_host + image_area * 1;
        float* phost_r = input_data_host + image_area * 2;
        /*
        float* phost_b = input_data_host + image_area * 3 * i + image_area * 0;
        float* phost_g = input_data_host + image_area * 3 * i + image_area * 1;
        float* phost_r = input_data_host + image_area * 3 * i + image_area * 2;
        */
        for(int i = 0; i < image_area; ++i, pimage += 3){
            *phost_r++ = (pimage[0] / 255.0f - mean[0]) / std[0];
            *phost_g++ = (pimage[1] / 255.0f - mean[1]) / std[1];
            *phost_b++ = (pimage[2] / 255.0f - mean[2]) / std[2];
        }

      //start2 = clock();
      ns_start2 = get_nanoseconds();
      iq.infer();
      ns_stop = get_nanoseconds();
      //stop = clock();
      //std::cout << "start2：" << start2 << std::endl;
      const int num_classes = 1000;
      //ClassificationResult classification_result(output, image_file_names, input_batch_size, 10, labels);
      ClassificationResult classification_result(output, {image_file_name}, input_batch_size, 10, labels);
      /*
      float* prob = output.data<float>();
      int predict_label = std::max_element(prob, prob + num_classes) - prob;

      float confidence = prob[predict_label];
      printf("confidence = %f, label_index = %d, label=%s\n", confidence, predict_label, labels[predict_label].c_str());
      */
      classification_result.show();
      //std::cout << "infer时间：" << ((double)(stop - start2)) / CLOCKS_PER_SEC << std::endl;
      //std::cout << "total时间：" << ((double)(stop - start1)) / CLOCKS_PER_SEC << std::endl;
      std::cout << "infer时间ns：" << ((double)(ns_stop - ns_start2)) / 1000000000.0 << std::endl;
      std::cout << "total时间ns：" << ((double)(ns_stop - ns_start1)) / 1000000000.0 << std::endl;

    }
}

int main(int argc, char *argv[]){
    if (argc != 2) {
        std::cout << "argc not 2: " << argc << std::endl;
        exit(-1);
    }

    std::vector<std::string> labels;
    boost::split(labels, std::string(argv[1]), boost::is_any_of(","));

    openvino_classifier(labels);
    return 0;
}

/*
classifier.onnx
./misc_classification_onnx_wthlabels car.jpeg
confidence = 0.228403, label_index = 817, label=sports car
infer时间ns：0.0497815
total时间ns：0.0505142

./misc_classification_onnx_wthlabels car.jpeg

Top 10 results:

Image car.jpeg

classid probability label
------- ----------- -----
817     0.2284026   sports car
472     0.1582138   canoe
784     0.1385197   screwdriver
693     0.0460029   paddle
590     0.0356732   hand-held computer
772     0.0245135   safety pin
867     0.0225798   trailer truck
511     0.0214936   convertible
473     0.0198144998   can opener
621     0.0198011   lawn mower

infer时间ns：0.05315462
total时间ns：0.05392876

./misc_classification_onnx_wthlabels car.jpeg,dog.jpeg,laptop.jpeg

Top 10 results:

Image car.jpeg

classid probability label
------- ----------- -----
817     0.2284025   sports car
472     0.1582143   canoe
784     0.1385197   screwdriver
693     0.0460029   paddle
590     0.0356731   hand-held computer
772     0.0245135   safety pin
867     0.0225798   trailer truck
511     0.0214935   convertible
473     0.0198998   can opener
621     0.0198011   lawn mower

Image dog.jpeg

classid probability label
------- ----------- -----
263     0.7865472   Pembroke
264     0.2117914   Cardigan
273     0.0001946   dingo
253     0.0001484   basenji
151     0.0001455   Chihuahua
231     0.0001416   collie
157     0.0001268   papillon
227     0.0000871   kelpie
232     0.0000750   Border collie
186     0.0000735   Norwich terrier

Image laptop.jpeg

classid probability label
------- ----------- -----
681     0.4685615   notebook
620     0.2356814   laptop
527     0.1043595   desktop computer
782     0.0676399   screen
810     0.0550855   space bar
590     0.0280587   hand-held computer
673     0.0074940   mouse
664     0.0066947   monitor
526     0.0042089   desk
916     0.0038441   web site

infer时间ns：0.0847067
total时间ns：0.08624184


resnet50.onnx
./misc_classification_onnx_wthlabels car.jpeg,dog.jpeg,laptop.jpeg

Top 10 resul1441ts:

Image car.jpeg

classid probability label
------- ----------- -----
817     9.1131859   sports car
472     8.7460270   canoe
784     8.6130886   screwdriver
693     7.5107803   paddle
590     7.2564735   hand-held computer
772     6.8812995   safety pin
867     6.7991319   trailer truck
511     6.7498288   convertible
473     6.6727839   can opener
621     6.6678138   lawn mower

Image dog.jpeg

classid probability label
------- ----------- -----
263     16.7579727  Pembroke
264     15.4459219  Cardigan
273     8.4533911   dingo
253     8.1822958   basenji
151     8.1629667   Chihuahua
231     8.1357012   collie
157     8.0254145   papillon
227     7.6495843   kelpie
232     7.4995999   Border collie
186     7.4799147   Norwich terrier

Image laptop.jpeg

classid probability label
------- ----------- -----
681     12.9896173  notebook
620     12.3024311  laptop
527     11.4877920  desktop computer
782     11.0541477  screen
810     10.8488359  space bar
590     10.1742477  hand-held computer
673     8.8540468   mouse
664     8.7412729   monitor
526     8.2771416   desk
916     8.1865005   web site

infer时间ns：0.06713036
total时间ns：0.06909315


resnet50-dynamic.onnx, 在3的batch_size里，把每个图片resize到第一个图片大小，预测结果都是错的
./misc_classification_onnx_wthlabels car.jpeg,dog.jpeg,laptop.jpeg

Top 10 results:

Image car.jpeg

classid probability label
------- ----------- -----
893     0.9895977   wallet
709     0.0049139   pencil box
446     0.0042856   binder
748     0.0005869   purse
593     0.0002521   harmonica
636     0.0000924   mailbag
696     0.0000720   paintbrush
592     0.0000248   hard disc
464     0.0000248   buckle
481     0.0000188   cassette

Image dog.jpeg

classid probability label
------- ----------- -----
814     0.7233258   speedboat
913     0.1111398   wreck
977     0.0739769   sandbar
144     0.0721574   pelican
978     0.0066975   seashore
472     0.0051869   canoe
693     0.0012136   paddle
975     0.0008178   lakeside
132     0.0004274   American egret
137     0.0004071   American coot

Image laptop.jpeg

classid probability label
------- ----------- -----
579     0.5943111   grand piano
881     0.4056732   upright
687     0.0000072   organ
553     0.0000017   file
526     0.0000012   desk
742     0.0000010   printer
401     0.0000010   accordion
446     0.0000008   binder
709     0.0000006   pencil box
770     0.0000004   running shoe

infer时间ns：0.05720701
total时间ns：0.05833848


resnet50-dynamic.onnx, 在1的batch_size里，不resize图片
image1.rows: 118
image1.cols: 122

Top 10 results:

Image car.jpeg

classid probability label
------- ----------- -----
893     0.9895977   wallet
709     0.0049139   pencil box
446     0.0042856   binder
748     0.0005869   purse
593     0.0002521   harmonica
636     0.0000924   mailbag
696     0.0000720   paintbrush
592     0.0000248   hard disc
464     0.0000248   buckle
481     0.0000188   cassette

infer时间ns：0.04185731
total时间ns：0.04237826
image1.rows: 118
image1.cols: 117

Top 10 results:

Image dog.jpeg

classid probability label
------- ----------- -----
169     0.7699625   borzoi
176     0.1981849   Saluki
172     0.0287884   whippet
231     0.0020972   collie
173     0.0003778   Ibizan hound
353     0.0002628   gazelle
690     0.0000573   oxcart
230     0.0000558   Shetland sheepdog
171     0.0000524   Italian greyhound
263     0.0000297   Pembroke

infer时间ns：0.01821995
total时间ns：0.06103708
image1.rows: 110
image1.cols: 120

Top 10 results:

Image laptop.jpeg

classid probability label
------- ----------- -----
791     0.5324421   shopping cart
790     0.3385024   shopping basket
746     0.0250895   puck
404     0.0180643   airliner
593     0.0174136   harmonica
628     0.0167843   liner
878     0.0155679   typewriter keyboard
754     0.0067585   radio
753     0.0051971   radiator
677     0.0041148   nail

infer时间ns：0.01679809
total时间ns：0.07846425


omz_downloader --name resnet-50-pytorch
omz_converter --name resnet-50-pytorch
#其实还是2步，先把pytorch转成onnx，onnx再转IR，导出onnx指定了input-shape

 */

File: openvino/misc_classification_onnx_wthlabels.py
"""
"""
import logging as log
import sys

from openvino.inference_engine import IECore
from openvino.runtime import Core
import cv2
import numpy as np

import time

def main():
    log.basicConfig(format='[ %(levelname)s ] %(message)s', level=log.INFO, stream=sys.stdout)

    if len(sys.argv) != 5:
        log.info(f'Usage: {sys.argv[0]} <path_to_model> <path_to_image> <device_name> <label_file_name>')
        return 1

    model_path = sys.argv[1]
    image_path = sys.argv[2]
    device_name = sys.argv[3]
    label_file_name = sys.argv[4]
    with open(label_file_name, 'r') as fp:
        labels = [line.replace('\n', '') for line in fp.readlines()]

    if model_path.endswith("-static.onnx"):
        log.info('Creating OpenVINO Runtime Core')
        ie = IECore()
        log.info(f'Reading the model:{model_path}')
        net = ie.read_network(model=model_path)
        exec_net = ie.load_network(network=net, device_name=device_name)
        input_blob = next(iter(net.input_info))
        output_blob = next(iter(net.outputs))
        _, _, h, w = net.input_info[input_blob].input_data.shape
    else:
        ie = Core()
        compiled_model = ie.compile_model(model=model_path, device_name=device_name)
        output_blob = compiled_model.output(0)
        h, w = 224, 224

    a1 = time.time()
    image_paths = image_path.split(',')
    images = []
    for image_path in image_paths:
        src = cv2.imread(image_path)
        sub_image = cv2.resize(src, (w, h))
        sub_image = cv2.cvtColor(sub_image, cv2.COLOR_BGR2RGB)
        sub_image = np.float32(sub_image) / 255.0
        sub_image[:, :, ] -= (np.float32(0.485), np.float32(0.456), np.float32(0.406),)
        sub_image[:, :, ] /= (np.float32(0.229), np.float32(0.224), np.float32(0.225),)
        sub_image = sub_image.transpose((2, 0, 1))
        images.append(sub_image)

    image = np.stack(images, axis=0)

    #log.info('Starting inference in synchronous mode')
    a2 = time.time()
    if model_path.endswith("-static.onnx"):
        #res = exec_net.infer(inputs={input_blob: [image]})
        res = exec_net.infer(inputs={input_blob: image})[output_blob]
    else:
        #image = np.expand_dims(image, 0)
        res = compiled_model([image])[output_blob]
    b = time.time()
    log.info(f'---debug---infer time:{b-a2}')
    log.info(f'---debug---preprocess + infer time:{b-a1}')

    for index, image_path in enumerate(image_paths):
        probs = res[index]
        top_10 = np.argsort(probs)[-10:][::-1]

        header = 'class_id        probability'

        log.info(f'Image path: {image_path}')
        log.info('Top 10 results: ')
        log.info(header)
        log.info('-' * len(header))

        for class_id in top_10:
            number_indent = len('class_id       ') - len(str(labels[class_id])) + 1
            probability_indent = ' ' * number_indent
            class_str = str(labels[class_id])
            log.info(f'{class_str}{probability_indent}{probs[class_id]:.7f}')

    log.info('')

    log.info('This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool\n')
    return 0

if __name__ == '__main__':
    sys.exit(main())


"""
/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-static.onnx car.jpeg CPU labels-imagenet.txt 
[ INFO ] Creating OpenVINO Runtime Core
[ INFO ] Reading the model:resnet50-static.onnx
[ INFO ] ---debug---infer time:0.09180426597595215
[ INFO ] ---debug---preprocess + infer time:0.27508020401000977
[ INFO ] Image path: car.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] SPORTS CAR      9.1131859
[ INFO ] CANOE           8.7460232
[ INFO ] SCREWDRIVER     8.6130886
[ INFO ] PADDLE          7.5107813
[ INFO ] HAND-HELD COMPUTER7.2564745
[ INFO ] SAFETY PIN      6.8813004
[ INFO ] TRAILER TRUCK   6.7991319
[ INFO ] CONVERTIBLE     6.7498298
[ INFO ] CAN OPENER      6.6727839
[ INFO ] LAWN MOWER      6.6678133
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-static.onnx laptop.jpeg CPU labels-imagenet.txt 
[ INFO ] Creating OpenVINO Runtime Core
[ INFO ] Reading the model:resnet50-static.onnx
[ INFO ] ---debug---infer time:0.09436917304992676
[ INFO ] ---debug---preprocess + infer time:0.11474394798278809
[ INFO ] Image path: laptop.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] NOTEBOOK        12.9896173
[ INFO ] LAPTOP          12.3024321
[ INFO ] DESKTOP COMPUTER11.4877920
[ INFO ] SCREEN          11.0541496
[ INFO ] SPACE BAR       10.8488340
[ INFO ] HAND-HELD COMPUTER10.1742487
[ INFO ] MOUSE           8.8540459
[ INFO ] MONITOR         8.7412739
[ INFO ] DESK            8.2771416
[ INFO ] WEB SITE        8.1865005
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-static.onnx dog.jpeg CPU labels-imagenet.txt 
[ INFO ] Creating OpenVINO Runtime Core
[ INFO ] Reading the model:resnet50-static.onnx
[ INFO ] ---debug---infer time:0.1067812442779541
[ INFO ] ---debug---preprocess + infer time:0.127913236618042
[ INFO ] Image path: dog.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] PEMBROKE        16.7579746
[ INFO ] CARDIGAN        15.4459257
[ INFO ] DINGO           8.4533911
[ INFO ] BASENJI         8.1822948
[ INFO ] CHIHUAHUA       8.1629667
[ INFO ] COLLIE          8.1357021
[ INFO ] PAPILLON        8.0254154
[ INFO ] KELPIE          7.6495848
[ INFO ] BORDER COLLIE   7.4995995
[ INFO ] NORWICH TERRIER 7.4799156
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-dynamic.onnx car.jpeg CPU labels-imagenet.txt 
[ INFO ] ---debug---infer time:0.18404603004455566
[ INFO ] ---debug---preprocess + infer time:0.18706202507019043
[ INFO ] Image path: car.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] SPORTS CAR      9.1131859
[ INFO ] CANOE           8.7460232
[ INFO ] SCREWDRIVER     8.6130886
[ INFO ] PADDLE          7.5107813
[ INFO ] HAND-HELD COMPUTER7.2564745
[ INFO ] SAFETY PIN      6.8813004
[ INFO ] TRAILER TRUCK   6.7991319
[ INFO ] CONVERTIBLE     6.7498298
[ INFO ] CAN OPENER      6.6727839
[ INFO ] LAWN MOWER      6.6678133
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-dynamic.onnx laptop.jpeg CPU labels-imagenet.txt 
[ INFO ] ---debug---infer time:0.14465093612670898
[ INFO ] ---debug---preprocess + infer time:0.16127467155456543
[ INFO ] Image path: laptop.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] NOTEBOOK        12.9896173
[ INFO ] LAPTOP          12.3024321
[ INFO ] DESKTOP COMPUTER11.4877920
[ INFO ] SCREEN          11.0541496
[ INFO ] SPACE BAR       10.8488340
[ INFO ] HAND-HELD COMPUTER10.1742487
[ INFO ] MOUSE           8.8540459
[ INFO ] MONITOR         8.7412739
[ INFO ] DESK            8.2771416
[ INFO ] WEB SITE        8.1865005
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-dynamic.onnx dog.jpeg CPU labels-imagenet.txt 
[ INFO ] ---debug---infer time:0.15482878684997559
[ INFO ] ---debug---preprocess + infer time:0.17514395713806152
[ INFO ] Image path: dog.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] PEMBROKE        16.7579746
[ INFO ] CARDIGAN        15.4459257
[ INFO ] DINGO           8.4533911
[ INFO ] BASENJI         8.1822948
[ INFO ] CHIHUAHUA       8.1629667
[ INFO ] COLLIE          8.1357021
[ INFO ] PAPILLON        8.0254154
[ INFO ] KELPIE          7.6495848
[ INFO ] BORDER COLLIE   7.4995995
[ INFO ] NORWICH TERRIER 7.4799156
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-dynamic.onnx car.jpeg,dog.jpeg CPU labels-imagenet.txt 
[ INFO ] ---debug---infer time:0.5754067897796631
[ INFO ] ---debug---preprocess + infer time:0.8114719390869141
[ INFO ] Image path: car.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] SPORTS CAR      9.1131868
[ INFO ] CANOE           8.7460279
[ INFO ] SCREWDRIVER     8.6130886
[ INFO ] PADDLE          7.5107803
[ INFO ] HAND-HELD COMPUTER7.2564735
[ INFO ] SAFETY PIN      6.8812990
[ INFO ] TRAILER TRUCK   6.7991319
[ INFO ] CONVERTIBLE     6.7498288
[ INFO ] CAN OPENER      6.6727829
[ INFO ] LAWN MOWER      6.6678143
[ INFO ] Image path: dog.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] PEMBROKE        16.7579746
[ INFO ] CARDIGAN        15.4459219
[ INFO ] DINGO           8.4533911
[ INFO ] BASENJI         8.1822968
[ INFO ] CHIHUAHUA       8.1629658
[ INFO ] COLLIE          8.1357012
[ INFO ] PAPILLON        8.0254145
[ INFO ] KELPIE          7.6495843
[ INFO ] BORDER COLLIE   7.4995999
[ INFO ] NORWICH TERRIER 7.4799147
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

/Volumes/data/envs/openvino/bin/python /Volumes/data/workspace/shouxieai/openvino/classification_onnx_wthlabels.py resnet50-dynamic.onnx car.jpeg,laptop.jpeg,dog.jpeg CPU labels-imagenet.txt 
[ INFO ] ---debug---infer time:0.3268129825592041
[ INFO ] ---debug---preprocess + infer time:0.3421509265899658
[ INFO ] Image path: car.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] SPORTS CAR      9.1131859
[ INFO ] CANOE           8.7460270
[ INFO ] SCREWDRIVER     8.6130886
[ INFO ] PADDLE          7.5107803
[ INFO ] HAND-HELD COMPUTER7.2564735
[ INFO ] SAFETY PIN      6.8812995
[ INFO ] TRAILER TRUCK   6.7991319
[ INFO ] CONVERTIBLE     6.7498288
[ INFO ] CAN OPENER      6.6727839
[ INFO ] LAWN MOWER      6.6678138
[ INFO ] Image path: laptop.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] NOTEBOOK        12.9896173
[ INFO ] LAPTOP          12.3024311
[ INFO ] DESKTOP COMPUTER11.4877920
[ INFO ] SCREEN          11.0541477
[ INFO ] SPACE BAR       10.8488359
[ INFO ] HAND-HELD COMPUTER10.1742477
[ INFO ] MOUSE           8.8540468
[ INFO ] MONITOR         8.7412729
[ INFO ] DESK            8.2771416
[ INFO ] WEB SITE        8.1865005
[ INFO ] Image path: dog.jpeg
[ INFO ] Top 10 results: 
[ INFO ] class_id        probability
[ INFO ] ---------------------------
[ INFO ] PEMBROKE        16.7579727
[ INFO ] CARDIGAN        15.4459219
[ INFO ] DINGO           8.4533911
[ INFO ] BASENJI         8.1822958
[ INFO ] CHIHUAHUA       8.1629667
[ INFO ] COLLIE          8.1357012
[ INFO ] PAPILLON        8.0254145
[ INFO ] KELPIE          7.6495843
[ INFO ] BORDER COLLIE   7.4995999
[ INFO ] NORWICH TERRIER 7.4799147
[ INFO ] 
[ INFO ] This sample is an API example, for any performance measurements please use the dedicated benchmark_app tool

"""